src/TreeView/useTypeahead.ts

```diff
@@ -0,0 +1,111 @@
+import React from 'react'
+
+type TypeaheadOptions = {
+  containerRef: React.RefObject<HTMLElement>
+  onFocusChange: (element: Element) => void
+}
+
+export function useTypeahead({containerRef, onFocusChange}: TypeaheadOptions) {
+  const [searchValue, setSearchValue] = React.useState('')
+  const timeoutRef = React.useRef(0)
+  const onFocusChangeRef = React.useRef(onFocusChange)
+
+  // Update the ref when the callback changes
+  React.useEffect(() => {
+    onFocusChangeRef.current = onFocusChange
+  }, [onFocusChange])
+
+  // Update the search value when the user types
+  React.useEffect(() => {
+    if (!containerRef.current) return
+    const container = containerRef.current
+
+    function onKeyDown(event: KeyboardEvent) {
+      // Ignore key presses that don't produce a character value
+      if (!event.key || event.key.length > 1) return
+
+      // Ignore key presses that occur with a modifier
+      if (event.ctrlKey || event.altKey || event.metaKey) return
+
+      // Update the existing search value with the new key press
+      setSearchValue(value => value + event.key)
+
+      // Reset the timeout
+      window.clearTimeout(timeoutRef.current)
+      timeoutRef.current = window.setTimeout(() => setSearchValue(''), 300)
```

Comment:
should this use `useSafeTimeout`?

Reply:
Ah, I didn't know about that hook. Great suggestion! Just updated it ✅